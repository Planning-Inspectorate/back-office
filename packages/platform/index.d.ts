import './config/dotenv';
import './http/ttl';
import './util/CryptoUtils';
import './util/date';
import './util/filter';
import './util/url';
import './validators/date';
import './validators/postcode';

declare module '@pins/platform' {
	export * from './config/dotenv';
	export * from './http/ttl';
	export * from './util/CryptoUtils';
	export * from './util/date';
	export * from './util/filter';
	export * from './util/url';
	export * from './validators/date';
	export * from './validators/postcode';
}
