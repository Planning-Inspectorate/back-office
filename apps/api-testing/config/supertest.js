// @ts-nocheck
import supertest from 'supertest';
import config from './config.cjs';
import { Logger } from '../utils/logger.js';

const logger = new Logger();
logger.log('green', '************* Tests starting ****************');
logger.log('cyan', `Running tests against: ${config.baseUrl}`);
logger.log('green', '*********************************************');

const request = supertest(config.baseUrl);

export default request;
