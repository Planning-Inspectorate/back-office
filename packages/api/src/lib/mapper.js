const mapQuestionnaireOutcome = (outcome) => {
  if (typeof outcome === 'undefined') {
    throw new Error('outcome is not defined');
  }

  switch (outcome) {
    case 'Complete':
      return 1;

    case 'Incomplete':
      return 2;

    default:
      return 3;
  }
};

module.exports = {
  mapQuestionnaireOutcome,
};
