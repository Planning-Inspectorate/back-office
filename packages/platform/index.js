export * from './http/hooks.js';
export * from './testing/assets/assets.js';
export * from './testing/fake.js';
export * from './testing/html-parser.js';
export * from './utils/date.js';
export * from './utils/dotenv.js';
export * from './utils/filter.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
