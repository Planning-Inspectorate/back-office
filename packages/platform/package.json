{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=16.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.0",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.1",
    "@types/lodash-es": "^4.17.6",
    "type-fest": "^2.12.2"
  }
}
