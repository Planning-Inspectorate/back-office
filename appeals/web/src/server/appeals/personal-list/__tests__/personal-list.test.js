import { parseHtml } from '@pins/platform';
import nock from 'nock';
import supertest from 'supertest';
import {
	assignedAppealsPage1,
	assignedAppealsPage2,
	assignedAppealsPage3
} from '#testing/app/fixtures/referencedata.js';
import { createTestEnvironment } from '#testing/index.js';
import { mapAppealStatusToActionRequiredHtml } from '../personal-list.mapper.js';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/personal-list';

describe('personal-list', () => {
	beforeEach(installMockApi);
	afterEach(teardown);

	describe('GET /', () => {
		it('should render the first page of the personal list with the expected content and pagination', async () => {
			nock('http://test/')
				.get('/appeals/my-appeals?pageNumber=1&pageSize=5')
				.reply(200, assignedAppealsPage1);

			const response = await request.get(`${baseUrl}${'?pageNumber=1&pageSize=5'}`);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});
		it('should render the second page of the personal list with the expected content and pagination', async () => {
			nock('http://test/')
				.get('/appeals/my-appeals?pageNumber=2&pageSize=5')
				.reply(200, assignedAppealsPage2);

			const response = await request.get(`${baseUrl}${'?pageNumber=2&pageSize=5'}`);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should render the second page of the personal list with applied filter, the expected content and pagination', async () => {
			nock('http://test/')
				.get('/appeals/my-appeals?pageNumber=2&pageSize=1&status=lpa_questionnaire_due')
				.reply(200, assignedAppealsPage3);

			const response = await request.get(
				`${baseUrl}${'?pageNumber=2&pageSize=1&appealStatusFilter=lpa_questionnaire_due'}`
			);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should render the header with navigation containing links to the personal list (with active modifier class), national list, and sign out route', async () => {
			nock('http://test/')
				.get('/appeals/my-appeals?pageNumber=1&pageSize=30')
				.reply(200, assignedAppealsPage1);

			const response = await request.get(baseUrl);
			const element = parseHtml(response.text, { rootElement: 'header' });

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should render a message when there are no cases assigned to the user', async () => {
			nock('http://test/')
				.get('/appeals/my-appeals?pageNumber=1&pageSize=5')
				.reply(200, { items: [], totalItems: 0, page: 1, totalPages: 1, pageSize: 5 });

			const response = await request.get(`${baseUrl}?pageNumber=1&pageSize=5`);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});
	});
});

describe('mapAppealStatusToActionRequiredHtml', () => {
	const appealId = 123;
	const lpaQuestionnaireId = 456;

	it('should return "Review appellant case" link for ready_to_start status with complete appellant case', () => {
		const result = mapAppealStatusToActionRequiredHtml(
			appealId,
			'ready_to_start',
			lpaQuestionnaireId,
			'Complete',
			'',
			'',
			false
		);
		expect(result).toEqual(
			`<a class="govuk-link" href="/appeals-service/appeal-details/${appealId}/appellant-case">Review appellant case</a>`
		);
	});

	it('should return "Awaiting appellant update" link for review_appellant_case status with incomplete appellant case', () => {
		const result = mapAppealStatusToActionRequiredHtml(
			appealId,
			'review_appellant_case',
			lpaQuestionnaireId,
			'Incomplete',
			'',
			'',
			false
		);
		expect(result).toEqual(
			`<a class="govuk-link" href="/appeals-service/appeal-details/${appealId}/appellant-case">Awaiting appellant update</a>`
		);
	});

	it('should return "Awaiting appellant update" text for review_appellant_case status with incomplete appellant case and isInspector true', () => {
		const result = mapAppealStatusToActionRequiredHtml(
			appealId,
			'review_appellant_case',
			lpaQuestionnaireId,
			'Incomplete',
			'',
			'',
			true
		);
		expect(result).toEqual('Awaiting appellant update');
	});

	it('should return "Awaiting LPA Questionnaire" for lpa_questionnaire_due status with no LPA Questionnaire ID', () => {
		const result = mapAppealStatusToActionRequiredHtml(
			appealId,
			'lpa_questionnaire_due',
			null,
			'',
			'',
			'',
			false
		);
		expect(result).toEqual('Awaiting LPA Questionnaire');
	});

	it('should return "Awaiting LPA update" link for lpa_questionnaire_due status with incomplete LPA Questionnaire', () => {
		const result = mapAppealStatusToActionRequiredHtml(
			appealId,
			'lpa_questionnaire_due',
			lpaQuestionnaireId,
			'',
			'Incomplete',
			'',
			false
		);
		expect(result).toEqual(
			`<a class="govuk-link" href="/appeals-service/appeal-details/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}">Awaiting LPA update</a>`
		);
	});

	it('should return "Awaiting LPA update" text for lpa_questionnaire_due status with incomplete LPA Questionnaire and isInspector true', () => {
		const result = mapAppealStatusToActionRequiredHtml(
			appealId,
			'lpa_questionnaire_due',
			lpaQuestionnaireId,
			'',
			'Incomplete',
			'',
			true
		);
		expect(result).toEqual('Awaiting LPA update');
	});

	it('should return "Review LPA Questionnaire" for lpa_questionnaire_due status with LPA Questionnaire', () => {
		const result = mapAppealStatusToActionRequiredHtml(
			appealId,
			'lpa_questionnaire_due',
			lpaQuestionnaireId,
			'',
			'',
			'',
			false
		);
		expect(result).toEqual(
			`<a class="govuk-link" href="/appeals-service/appeal-details/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}">Review LPA Questionnaire</a>`
		);
	});

	it('should return "Review LPA Questionnaire" for lpa_questionnaire_due status with LPA Questionnaire and isInspector true', () => {
		const result = mapAppealStatusToActionRequiredHtml(
			appealId,
			'lpa_questionnaire_due',
			lpaQuestionnaireId,
			'',
			'',
			'',
			true
		);
		expect(result).toEqual('Review LPA Questionnaire');
	});

	it('should return "LPA Questionnaire Overdue" for lpa_questionnaire_due status with LPA Questionnaire overdue', () => {
		const result = mapAppealStatusToActionRequiredHtml(
			appealId,
			'lpa_questionnaire_due',
			null,
			'',
			'',
			'2024-01-01',
			false
		);
		expect(result).toEqual('LPA Questionnaire Overdue');
	});

	it('should return "Submit decision" link for issue_determination status', () => {
		const result = mapAppealStatusToActionRequiredHtml(
			appealId,
			'issue_determination',
			null,
			'',
			'',
			'',
			false
		);
		expect(result).toEqual(
			`<a class="govuk-link" href="/appeals-service/appeal-details/${appealId}/issue-decision/decision">Submit decision</a>`
		);
	});

	it('should return "Update Horizon reference" link for awaiting_transfer status', () => {
		const result = mapAppealStatusToActionRequiredHtml(
			appealId,
			'awaiting_transfer',
			null,
			'',
			'',
			'',
			false
		);
		expect(result).toEqual(
			`<a class="govuk-link" href="/appeals-service/appeal-details/${appealId}/change-appeal-type/add-horizon-reference">Update Horizon reference</a>`
		);
	});

	it('should return "View appellant case" link for any other status', () => {
		const result = mapAppealStatusToActionRequiredHtml(
			appealId,
			'some_other_status',
			null,
			'',
			'',
			'',
			false
		);
		expect(result).toEqual(
			`<a class="govuk-link" href="/appeals-service/appeal-details/${appealId}">View appellant case</a>`
		);
	});
});
