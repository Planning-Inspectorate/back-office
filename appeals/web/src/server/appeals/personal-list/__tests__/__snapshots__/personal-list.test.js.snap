// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`personal-list GET / should render the personal list with the expected content 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-full">
            <h1 class="govuk-heading-l govuk-!-margin-bottom-6">Cases assigned to you</h1>
        </div>
    </div>
    <table class="govuk-table">
        <thead class="govuk-table__head">
            <tr class="govuk-table__row">
                <th scope="col" class="govuk-table__header">Appeal ID</th>
                <th scope="col" class="govuk-table__header">Lead or child</th>
                <th scope="col" class="govuk-table__header">Action required</th>
                <th scope="col" class="govuk-table__header">Due by</th>
                <th scope="col" class="govuk-table__header">Case status</th>
            </tr>
        </thead>
        <tbody class="govuk-table__body">
            <tr class="govuk-table__row">
                <td class="govuk-table__cell"><strong><a class="govuk-link" href="/appeals-service/appeal-details/132" aria-label="Appeal 2 6 8 4 7 0">268470</a></strong>
                </td>
                <td class="govuk-table__cell"><strong class="govuk-tag govuk-tag--grey single-line"></strong>
                </td>
                <td class="govuk-table__cell"><a class="govuk-link" href="/appeals-service/appeal-details/132/appellant-case">View appellant case</a>
                </td>
                <td class="govuk-table__cell">6 December 2023</td>
                <td class="govuk-table__cell"><strong class="govuk-tag govuk-tag--turquoise single-line">ready to start</strong>
                </td>
            </tr>
            <tr class="govuk-table__row">
                <td class="govuk-table__cell"><strong><a class="govuk-link" href="/appeals-service/appeal-details/147" aria-label="Appeal 6 5 2 8 9 6">652896</a></strong>
                </td>
                <td class="govuk-table__cell"><strong class="govuk-tag govuk-tag--grey single-line"></strong>
                </td>
                <td class="govuk-table__cell"><a class="govuk-link" href="/appeals-service/appeal-details/147/lpa-questionnaire/59">View LPA Questionnaire</a>
                </td>
                <td class="govuk-table__cell">8 December 2023</td>
                <td class="govuk-table__cell"><strong class="govuk-tag govuk-tag--yellow single-line">lpa questionnaire due</strong>
                </td>
            </tr>
            <tr class="govuk-table__row">
                <td class="govuk-table__cell"><strong><a class="govuk-link" href="/appeals-service/appeal-details/154" aria-label="Appeal 5 0 1 6 5">50165</a></strong>
                </td>
                <td class="govuk-table__cell"><strong class="govuk-tag govuk-tag--grey single-line"></strong>
                </td>
                <td class="govuk-table__cell"><a class="govuk-link" href="/appeals-service/appeal-details/154">View appeal</a>
                </td>
                <td class="govuk-table__cell">25 January 2024</td>
                <td class="govuk-table__cell"><strong class="govuk-tag govuk-tag--orange single-line">statement review</strong>
                </td>
            </tr>
        </tbody>
    </table>
</main>"
`;
